README.md
![flaiir](/images/banner.png)
An AI tool that utilizes the OpenAI API to generate code suggestions, fixes spelling mistakes, refactor existing code and explains code for the user, based on the current text editors selection and file type.

---

## Features

- Generate code from a highlighted comment/code.
  - Shown inline under the highlighted comment/code.
- Explain code.
- Refactor code.
- Auto detecting your current file extension (.js, .ts, .py etc..) to deliver the code in the correct format.
- Quickly fix spelling mistakes in comments, Markdown and text files.

---

## **Requirements**

- Create a free account on [https://beta.openai.com/](https://beta.openai.com/), you will start with 18$ in free credit that can be used during your first 3 months.

- Go to flaiir workspace settings
  ![settings](/images/settings.png)
- Your API Key, can be found here: [https://beta.openai.com/account/api-keys](https://beta.openai.com/account/api-keys)
- Organization ID can be found here [https://beta.openai.com/account/org-settings](https://beta.openai.com/account/org-settings)
- The models you wish to use, the models can be found here [https://beta.openai.com/docs/models/gpt-3](https://beta.openai.com/docs/models/gpt-3)
- Temperature, a value between 0-1.

---

## How to use

- Install extension

## TL;DR

```txt
Highlight the comment/code you want to generate/explain/refactor some code or text from.

Open command palette:

shift + command + p

Commands:

flaiir generator/explainer
	* Generates code and explains code

flaiir refactor
  *  This command refactors code and fixes spelling mistakes
```

---

### Generate code

- Highlight the comment/code your want to generate code from

For example, write the following comment in a file and hightlight it with your mouse.

Input:

```js
// create a fetch function with async/await
```

Then

- Open command palette: **shift + command + p** and type “flaiir generator/explainer” and hit the _Enter key_

- Wait for the code to be generated...

Output:

```js
const fetchData = async url => {
	try {
		const response = await fetch(url);
		const data = await response.json();
		return data;
	} catch (error) {
		console.log(error);
	}
};
```

---

### Refactor

Input:

```js
//change fetchdata to use then instead of async/await
const fetchData = async url => {
	try {
		const response = await fetch(url);
		const data = await response.json();
		return data;
	} catch (error) {
		console.log(error);
	}
};
```

- Open command palette: **shift + command + p** and type “flaiir refactor” and hit the _Enter key_

Output:

```js
const fetchData = url => {
	return fetch(url)
		.then(response => response.json())
		.then(data => data)
		.catch(error => console.log(error));
};
```

---

### Explain code

Input:

```js
// Function 1
var fullNames = [];
for (var i = 0; i < 50; i++) {
	fullNames.push(
		names[Math.floor(Math.random() * names.length)] + ' ' + lastNames[Math.floor(Math.random() * lastNames.length)]
	);
}

// What does Function 1 do?
```

- Highlight all of the above
  - Open command palette: **shift + command + p** and type “flaiir generator/explainer” and hit the _Enter key_

Output:

```js
// Function 1 creates an array of 50 random full names.
```

---

### Fix spelling mistakes

- Provide a one-line instruction, then highlight the instruction and paragraph you want to fix.

Input

```md
fix spelling mistakes

At the start of school Dora was afrad of her new Teacher. Mrs. Davis seamed nice, but she had so manny rules for the class to folow. Scare someone to pieces. As the school year cotinued, Dora begun to understan how the Teacher come up with the rules The rules were their so students would be respecful of theyselves and each other. By the end of the year, Dora though Mrs. Davis was the best Teacher she evere had!
```

Then

- Open command palette: **shift + command + p** and type “flaiir refactor” and hit the _Enter key_

Output

```md
At the start of school Dora was afraid of her new Teacher. Mrs. Davis seemed nice, but she had so many rules for the class to follow. Scare someone to pieces. As the school year continued, Dora began to understand how the Teacher come up with the rules The rules were their so students would be respectful of them selves and each other. By the end of the year, Dora thought Mrs. Davis was the best Teacher she ever had!
```

---


CHANGELOG.md
# Change Log

All notable changes to the "flaiir" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

## [1.0.0] - 2023-15-01

### Changed

- Using flaiir.config.json to using vs code workspace settings

## [1.0.1] - 2023-15-01

### Added

- Loading indicators


src/code_edit.ts
import * as vscode from 'vscode';
import { Configuration, OpenAIApi } from 'openai';
import * as path from 'path';

async function getCodeEdit(): Promise<void> {
	const vsconfig = vscode.workspace.getConfiguration();
	let openAI_apiKey = vsconfig.get('openai.apiKey') as string;
	let openAI_organizationID = vsconfig.get('openai.organizationID') as string;
	let openAI_edit_model = vsconfig.get('openai.edit.model') as string;
	let openAI_edit_temperature = vsconfig.get('openai.edit.temperature') as number;

	if (!openAI_apiKey || !openAI_organizationID) {
		vscode.window.showErrorMessage('Please check your settings, apiKey and organizationID are required.');
		return;
	}
	// Set up the configuration object
	const configuration = new Configuration({
		organization: openAI_organizationID,
		apiKey: openAI_apiKey,
	});

	// Create the OpenAIApi object
	const openai = new OpenAIApi(configuration);
	// Set the model to use for code completions
	const edit_model = openAI_edit_model;
	const edit_temperature = openAI_edit_temperature;

	// Get the current text editor
	let editor = vscode.window.activeTextEditor;
	if (editor) {
		// Get the position of the last line of the highlighted text
		let selection = editor.selection;

		//get filepath from the currently open file
		const filePath = editor?.document.uri.fsPath;
		let fileExtension;
		if (filePath) fileExtension = path.extname(filePath);

		// Modify the prompt to include the cursor position
		const instruction = editor.document.lineAt(selection.start.line).text;
		const input = editor.document.getText(
			new vscode.Range(selection.start.line + 1, 0, selection.end.line, selection.end.character)
		);
		// Get the code edit suggestions
		try {
			const response = await openai.createEdit({
				model: edit_model ? edit_model : 'code-davinci-edit-001',
				input,
				instruction,
				temperature: edit_temperature ? edit_temperature : 0.6,
				top_p: 1,
			});

			// Replace the highlighted text with the suggestions
			editor.edit(editBuilder => {
				editBuilder.replace(selection, response.data.choices[0].text as string);
			});
		} catch (error: any) {
			if (error.response.status === 401) {
				vscode.window.showErrorMessage('The API key and organization ID are not correct or invalid.');
			} else {
				vscode.window.showErrorMessage(error.message);
			}
		}
	} else {
		vscode.window.showErrorMessage('No text editor is active. Please open a file first.');
	}
}
const editCode = vscode.commands.registerCommand('flaiir.getCodeEdit', async () => {
	try {
		await vscode.window.withProgress(
			{
				location: vscode.ProgressLocation.Notification,
				title: 'Loading...',
				cancellable: false,
			},
			async (progress, token) => {
				token.onCancellationRequested(() => {
					console.log('User canceled the long running operation');
				});
				await getCodeEdit();
				progress.report({ increment: 100 });
			}
		);
	} catch (err: any) {
		if (err.code === 'ENOENT') {
			vscode.window.showErrorMessage('settings.json file not found.');
		} else if (err instanceof SyntaxError) {
			vscode.window.showErrorMessage('settings.json file is not a valid JSON file.');
		} else {
			vscode.window.showErrorMessage('An error occurred while reading settings.json file.');
		}
	}
});

export default editCode;


src/code_webview.ts
import * as vscode from 'vscode';
import { Configuration, OpenAIApi } from 'openai';
import path = require('path');

async function getCodeSuggetions(jsonContent: any): Promise<void> {
	// Set up the configuration object
	const configuration = new Configuration({
		organization: jsonContent.organization,
		apiKey: jsonContent.apiKey,
	});

	// Create the OpenAIApi object
	const openai = new OpenAIApi(configuration);
	// Set the model to use for code completions
	const model = jsonContent.model;

	// Set the prompt to use for code completions
	// Get the current text editor
	let decorationType = vscode.window.createTextEditorDecorationType({
		overviewRulerLane: vscode.OverviewRulerLane.Right,
		before: {
			margin: '-1px',
			contentText: '',
			border: 'solid 2px #28efa8',
		},
	});

	let editor = vscode.window.activeTextEditor;
	if (editor) {
		let selection = editor.selection;
		let decorations = [{ range: selection }];
		console.log(decorations);
		editor.setDecorations(decorationType, decorations);
	}

	//get filepath from the currently open file
	const filePath = editor?.document.uri.fsPath;
	let fileExtension;
	if (filePath) fileExtension = path.extname(filePath);

	// Get the selected text
	const selection = editor?.selection;
	const code_snippet = editor?.document.getText(selection);

	// Get the cursor position
	const cursor_position = selection?.active.character;

	// Modify the prompt to include the cursor position
	const prompt =
		code_snippet?.substring(0, cursor_position) +
		'|' +
		code_snippet?.substring(cursor_position as number) +
		'in' +
		fileExtension;
	console.log(prompt, 'prompt');

	// Get the code completion suggestions
	const response = await openai.createCompletion({
		model,
		prompt,
		max_tokens: 1024,
		temperature: 0,
		top_p: 1,
		frequency_penalty: 0,
		presence_penalty: 0,
		best_of: 1,
		n: 1,
	});

	// ** Use Vs code webview **
	const panel = vscode.window.createWebviewPanel('getCodeSuggestions', 'Code suggestions', vscode.ViewColumn.One, {
		enableScripts: true, // Allows scripts in the webview
	});

	panel.webview.html = getWebviewContent();

	function getWebviewContent() {
		return `<!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Code Suggestions</title>
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.7.0/styles/atom-one-dark.min.css">
    </head>
    <body>
			<header style="text-align: center; padding: 10px;">
				<h1>Here is the requested code</h1>
				<p style="font-size: smaller;">Click anywhere on the code to copy</p>
			</header>
        <pre>
            <code id="code-element" >
            ${response.data.choices[0].text}		
            </code>
        </pre>
        <script>
            const codeElement = document.getElementById('code-element');
            codeElement.addEventListener('click', () => {
                navigator.clipboard.writeText(codeElement.textContent);
            });
        </script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.7.0/highlight.min.js" integrity="sha512-bgHRAiTjGrzHzLyKOnpFvaEpGzJet3z4tZnXGjpsCcqOnAH6VGUx9frc5bcIhKTVLEiCO6vEhNAgx5jtLUYrfA==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
        <script>hljs.highlightAll();</script>
    </body>
    </html>`;
	}
}

export default getCodeSuggetions;


src/code_suggestions.ts
import * as vscode from 'vscode';
import { Configuration, OpenAIApi } from 'openai';
import * as path from 'path';

async function generateCode(): Promise<void> {
	const vsconfig = vscode.workspace.getConfiguration();
	let openAI_apiKey = vsconfig.get('openai.apiKey') as string;
	let openAI_organizationID = vsconfig.get('openai.organizationID') as string;
	let openAI_completion_model = vsconfig.get('openai.completion.model') as string;
	let completion_temperature = vsconfig.get('openai.completion.temperature') as number;
	if (!openAI_apiKey || !openAI_organizationID) {
		vscode.window.showErrorMessage('Please check your settings, apiKey and organizationID are required.');
		return;
	}

	// Set up the configuration object
	const configuration = new Configuration({
		organization: openAI_organizationID,
		apiKey: openAI_apiKey,
	});

	// Create the OpenAIApi object
	const openai = new OpenAIApi(configuration);

	// Set the model to use for code completions
	const completionModel = openAI_completion_model;

	// Get the current text editor
	let editor = vscode.window.activeTextEditor;
	if (editor) {
		// Get the position of the last line of the highlighted text
		let selection = editor.selection;
		let lastLine = selection.end.line;
		// Get the last character of the last line
		let lastChar = editor.document.lineAt(lastLine).range.end;

		//get filepath from the currently open file
		const filePath = editor?.document.uri.fsPath;
		let fileExtension;
		if (filePath) fileExtension = path.extname(filePath);

		// Get the selected text
		const code_snippet = editor?.document.getText(selection);

		// Get the cursor position
		const cursor_position = selection?.active.character;

		// Modify the prompt to include the cursor position
		const prompt =
			code_snippet?.substring(0, cursor_position) +
			'|' +
			code_snippet?.substring(cursor_position as number) +
			'in' +
			fileExtension;

		// Get the code completion suggestions
		try {
			const response = await openai.createCompletion({
				model: completionModel ? completionModel : 'text-davinci-003',
				prompt,
				max_tokens: 1024,
				temperature: completion_temperature ? completion_temperature : 0,
				top_p: 1,
				frequency_penalty: 0,
				presence_penalty: 0,
				best_of: 1,
				n: 1,
			});

			// Insert the suggestions after the last character of the last line
			editor.edit(editBuilder => {
				editBuilder.insert(lastChar, '\n' + response.data.choices[0].text);
			});
		} catch (error: any) {
			if (error.response.status === 401) {
				vscode.window.showErrorMessage('The API key and organization ID are not correct or invalid.');
			} else {
				vscode.window.showErrorMessage(error.message);
			}
		}
	} else {
		vscode.window.showErrorMessage('No text editor is active. Please open a file first.');
	}
}

const getCodeSuggestions = vscode.commands.registerCommand('flaiir.getCodeSuggestions', async () => {
	try {
		await vscode.window.withProgress(
			{
				location: vscode.ProgressLocation.Notification,
				title: 'Loading code suggestions...',
				cancellable: false,
			},
			async (progress, token) => {
				token.onCancellationRequested(() => {
					console.log('User canceled the long running operation');
				});
				await generateCode();
				progress.report({ increment: 100 });
			}
		);
	} catch (err: any) {
		if (err.code === 'ENOENT') {
			vscode.window.showErrorMessage('settings.json file not found.');
		} else if (err instanceof SyntaxError) {
			vscode.window.showErrorMessage('settings.json file is not a valid JSON file.');
		} else {
			vscode.window.showErrorMessage('An error occurred while reading settings.json file.');
		}
	}
});

export default getCodeSuggestions;
